### PR TITLE
feat(AlgebraicGeometry/ProjectiveSpectrum/Scheme): add the underlying map of forward  sheaf component

### PR DESCRIPTION
**Pull request by jjaassoonn**
for each open sets `V ⊆ Spec A⁰_f`, We aim to define `(Spec A⁰_f)(V) ⟶ (Proj| pbo f)(φ⁻¹ V)`. Let `s` be a section on `V` (i.e. a (dependent) function `Π (𝓅 ∈ V) (A⁰_f)_𝓅`) and `y : φ⁻¹ V`, we consider the fraction `s φ(y) ∈ (A⁰_f)_φ(y)`, we can write it as `a/b` where `a, b` are fractions in `A⁰_f`; by writing `a` as `n_a/m_a` and `b` as `n_b/m_b`, we obtain a homogeneous fraction in `A⁰_y`

```
              n_a * m_b
    α(s)(y) = ---------.
              n_b * m_a
```

Note that this is just a function, still need to prove that it preserves zero, one, addition, multiplication, and locally a fraction

Should be replaced by #12371

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by jjaassoonn** [comment]
!bench

File: Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean [hunk 1]
```diff
@@ -57,6 +57,22 @@ open sets in `Proj`, more specifically:
     to prove `fromSpec ∘ toSpec` and `toSpec ∘ fromSpec` are both identities; these are achieved in
     `ProjIsoSpecTopComponent.fromSpec_toSpec` and `ProjIsoSpecTopComponent.toSpec_fromSpec`.
 
+    `ProjIsoSpecTopComponent.FromSpec.toFun`. But to prove the continuity of this function, we need
+    to prove `fromSpec ∘ toSpec` and `toSpec ∘ fromSpec` are both identities; these are achieved in
+    `ProjIsoSpecTopComponent.fromSpecToSpec` and `ProjIsoSpecTopComponent.toSpecFromSpec`.
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean:62]: I think these are duplicated.

Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/ProjectiveSpectrum/Scheme.lean]: Thanks. Removed

**Comment by jjaassoonn** [comment]
> I haven't looked at the whole file so I might be totally wrong, but couldn't `(Spec A⁰_f)(V) -> (φ _* Proj | D(f))(V)` be obtained easily via the Gamma-spec adjunction and a the ring map `Proj(D(f)) -> A⁰_f` used to construct the topological maps?

Ah I completely forgot about gamma spec adjunction. But I don't think I have a map `Proj(D(f)) -> A⁰_f`.

Do you think gamma spec adjunction would work for the backward direction as well? Because this very explicit approach (forward and backward direction) has the advantage of making checking they are inverse to each other just about calculating fractions.

**Comment by erdOne** [comment]
It's a typo. It should be `A⁰_f -> Proj(D(f))` instead, which maps x to the constant function with value x. Then just check that the sheaf map is bijective on stalks, which should be easy on paper (but maybe not in lean).

**Comment by jjaassoonn** [comment]
> It's a typo. It should be `A⁰_f -> Proj(D(f))` instead, which maps x to the constant function with value x. Then just check that the sheaf map is bijective on stalks, which should be easy on paper (but maybe not in lean).

Do we have the fact sheaf morphism is an isomorphism if and only if the induced stalk ring homs are isomorphisms in Lean?

**Comment by erdOne** [comment]
We have `TopCat.Presheaf.app_bijective_of_stalkFunctor_map_bijective` or `TopCat.Presheaf.isIso_iff_stalkFunctor_map_iso`.

**Comment by jjaassoonn** [comment]
Thanks. I will try tomorrow the gamma spec adjunction approach.

**Comment by jjaassoonn** [comment]
I think this  is how gamma-spec adjunction going to work:

We define the following sheaf morphism
```lean
example : (Proj| (pbo f)) ⟶ (Spec (A⁰_ f)) :=
  ΓSpec.locallyRingedSpaceAdjunction.homEquiv (Proj| (pbo f)) (op <| .of <| A⁰_ f) <| op <|
    {
      toFun := fun x ↦
        ⟨fun y ↦ Quotient.mk'' ⟨x.deg, ⟨x.num, x.num_mem_deg⟩, ⟨x.den, x.den_mem_deg⟩,
          by
            simp only [Functor.op_obj, unop_op, LocallyRingedSpace.forgetToSheafedSpace_obj,
              SheafedSpace.forgetToPresheafedSpace_obj, LocallyRingedSpace.restrict_carrier,
              IsOpenMap.functor_obj_coe, Opens.coe_inclusion, Opens.coe_top]
            obtain ⟨n, (hn : f^n = x.den)⟩ := x.den_mem
            intro r
            rw [← hn] at r
            have y_mem := y.2
            simp only [Functor.op_obj, unop_op, LocallyRingedSpace.forgetToSheafedSpace_obj,
              SheafedSpace.forgetToPresheafedSpace_obj, LocallyRingedSpace.restrict_carrier,
              Opens.openEmbedding_obj_top] at y_mem
            exact y_mem (y.1.2.mem_of_pow_mem _ r)⟩, by
          rintro ⟨z, hz⟩
          simp only [Functor.op_obj, unop_op, LocallyRingedSpace.forgetToSheafedSpace_obj,
            SheafedSpace.forgetToPresheafedSpace_obj, LocallyRingedSpace.restrict_carrier,
            Opens.openEmbedding_obj_top] at hz
          refine ⟨_, hz, eqToHom (by ext; simp), x.deg, ⟨x.num, x.num_mem_deg⟩,
            ⟨x.den, x.den_mem_deg⟩, fun a ↦ ⟨?_, rfl⟩⟩
          obtain ⟨n, (hn : f^n = x.den)⟩ := x.den_mem
          rintro (r : x.den ∈ _)
          rw [← hn] at r
          exact a.2 <| a.1.2.mem_of_pow_mem _ r⟩
      map_one' := by
        simp only [Functor.rightOp_obj, LocallyRingedSpace.Γ_obj, unop_op,
          LocallyRingedSpace.restrict_carrier, LocallyRingedSpace.restrict_presheaf_obj,
          Functor.op_obj, LocallyRingedSpace.forgetToSheafedSpace_obj,
          SheafedSpace.forgetToPresheafedSpace_obj, IsOpenMap.functor_obj_coe, Opens.coe_inclusion,
          Opens.coe_top, CommRingCat.coe_of]
        refine Subtype.ext <| funext fun x ↦ ?_
        simp only [Functor.op_obj, unop_op, LocallyRingedSpace.forgetToSheafedSpace_obj,
          SheafedSpace.forgetToPresheafedSpace_obj, LocallyRingedSpace.restrict_carrier,
          IsOpenMap.functor_obj_coe, Opens.coe_inclusion, Opens.coe_top, CommRingCat.coe_of]
        change _ = 1

        rw [HomogeneousLocalization.ext_iff_val, HomogeneousLocalization.one_val,
          HomogeneousLocalization.val_mk'',
          show (1 : Localization.AtPrime x.1.asHomogeneousIdeal.toIdeal) = Localization.mk 1 1 by
            convert Localization.mk_self _ |>.symm
            rfl]

        erw [Localization.mk_eq_mk_iff, Localization.r_iff_exists]
        have eq_final := (1 : A⁰_f).eq_num_div_den
        rw [HomogeneousLocalization.one_val,
          show (1 : Localization.Away f) = Localization.mk 1 1 by
            convert Localization.mk_self _ |>.symm
            rfl, Localization.mk_eq_mk_iff, Localization.r_iff_exists] at eq_final
        obtain ⟨⟨_, ⟨n, rfl⟩⟩, eq_final⟩ := eq_final
        simp only [mul_one, OneMemClass.coe_one, one_mul] at eq_final
        refine ⟨⟨f^n, ?_⟩, by simpa using eq_final.symm⟩
        have mem_final := x.2
        simp only [Functor.op_obj, unop_op, LocallyRingedSpace.forgetToSheafedSpace_obj,
          SheafedSpace.forgetToPresheafedSpace_obj, LocallyRingedSpace.restrict_carrier,
          Opens.openEmbedding_obj_top] at mem_final
        exact fun r ↦ mem_final <| x.1.2.mem_of_pow_mem _ r
      map_mul' := sorry
      map_zero' := sorry
      map_add' := sorry
    }
```

and we prove that the induced stalk map is isomorphism. But I am not fully convinced that I want to make the switch. Because sending $x$ to the constant function $x$ is still kinda messy: $x$ is in $A^0_f$ and the "constant function x" is in $A^0_{\mathfrak p}$ for some homogeneous ideal $\mathfrak{p} \in D(f)$, so coercion is needed as well. So works still needs to be done to prove preservation of zero/one/addition/multiplication, being locally fraction. Maybe its less work compared to this approach though.

Checking the induced stalk map is isomorphism probably won't be very easy to formalize as well.

**Comment by jjaassoonn** [comment]
> I haven't looked at the whole file so I might be totally wrong, but couldn't `(Spec A⁰_f)(V) -> (φ _* Proj | D(f))(V)` be obtained easily via the Gamma-spec adjunction and a the ring map `Proj(D(f)) -> A⁰_f` used to construct the topological maps?

Using Gamma-Spec adjunction would indeed be extremely elegant and even make the homeomorphism unnecessary I think.

Is this the argument using Gamma-Spec adjunction:

Let $\alpha: A^0_f \to \mathrm{Proj}|_{D(f)}$ be function defined by sending $$x$$ to the constant function $x$. Then the morphism from $\mathrm{Proj}|_{D(f)} \to \mathrm{Spec}(A^0_f)$ is defined by
```math
\mathrm{Proj}|_{D(f)} \to \mathrm{Spec}(\mathrm{Proj}_{D(f)}) \stackrel{\mathrm{Spec(\alpha)}}{\to} \mathrm{Spec}(A^0_f)
```
Then we need to check the induced stalk morphism at $\mathfrak q \in \mathrm{Proj}|_(D(f))$ is bijective, i.e. 
```math
{(A^0_f)}_{\mathfrak p} \to A^0_{\mathfrak q}
```
is bijective where $\mathfrak p$ is calculated as the preimage of the maximal ideal of $A^0_{\mathfrak q}$ under 
```math
A^0_f \to \Gamma(\mathcal{O}_{\mathrm{Proj}}, D(f)) \to \mathcal{O}_{\mathrm{Proj}, \mathfrak q} \cong A^0_{\mathfrak q}.
```

Ideally $\mathfrak p$ should just be $A^0_f \cap \mathrm{span} \{\frac{g}{1} | g \in \mathfrak q\}$ (this should be okay since $f \not\in \mathfrak q$.)  and
${(A^0_f)}_{\mathfrak p} \to A^0_{\mathfrak q}$ should ideally have inverse $A^0_{\mathfrak q}\ni \frac{a}{b} \mapsto \frac{a/1}{b/1} \in {(A^0_f)}_{\mathfrak p}$. (I am not quite sure what the forward ring map is, but I am convinced that this is its inverse)

But since $\mathcal{O}_{\mathrm{Proj}, \mathfrak q} \cong A^0_{\mathfrak q}$ and $\mathcal{O}_{\mathrm{Spec(A^0_f)}, \mathfrak p} \cong {(A^0_f)}_{\mathfrak p}$ are both complicated in Lean, I am not sure how easy it is to check that the stalk map is isomorphism.

**Comment by erdOne** [comment]
Sorry forgot to reply.
I don't think this isomorphism is a one off thing. We will need to identify $D_+(f)$ with $\mathrm{Spec}(A^0_f)$ at times and I think it might be easier to show that this isomorphism commutes with other stuff if this map comes from some general machinery and so that the API of the machinery can be used. Maybe this don't matter though as the adjunction guarantees that the map is uniquely determined. 

If you have some more thoughts, maybe we should move this conversation to Zulip?

**Comment by jjaassoonn** [comment]
> Sorry forgot to reply. I don't think this isomorphism is a one off thing. We will need to identify D+(f) with Spec(Af0) at times and I think it might be easier to show that this isomorphism commutes with other stuff if this map comes from some general machinery and so that the API of the machinery can be used. Maybe this don't matter though as the adjunction guarantees that the map is uniquely determined.
> 
> If you have some more thoughts, maybe we should move this conversation to Zulip?

The thread is [here](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2311920.20towards.20proj.20construction)

**Comment by jjaassoonn** [comment]
Superceded

**Pull request closed without merging**